runtime/j9vm/java9vmi.c

```diff
@@ -951,93 +952,87 @@ JVM_DefineModule(JNIEnv * env, jobject module, jstring version, jstring location
 			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module is unnamed");
 		} else if (!isModuleNameValid(moduleName)) {
 			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module name is invalid");
-		} else if (NULL != classLoader) {
-			BOOLEAN success = FALSE;
-			J9Module *j9mod = createModule(currentThread, modObj, classLoader, moduleName);
-			UDATA rc = ERRCODE_GENERAL_FAILURE;
-
-			if (NULL != j9mod) {
+		} else if (NULL == classLoader) {
+			/* An exception should be pending if classLoader is null */
+			assert(NULL != currentThread->currentException);
+		} else {
+			char buf[J9VM_PACKAGE_NAME_BUFFER_LENGTH];
+			char *nameUTF = buf;
+
+			PORT_ACCESS_FROM_VMC(currentThread);
+			nameUTF = vmFuncs->copyStringToUTF8WithMemAlloc(
+				currentThread, moduleName, J9_STR_NULL_TERMINATE_RESULT, "", 0, buf, J9VM_PACKAGE_NAME_BUFFER_LENGTH, NULL);
+			if (NULL == nameUTF) {
+				vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			} else if ((0 == strcmp(nameUTF, JAVA_BASE_MODULE)) && (classLoader != vm->systemClassLoader)) {
```

Comment:
The classLoader check is cheaper as it's a pointer comparison.  We should probably do it first.

Reply:
Fixed.